android/buildAndroidBOINC-CI.sh

```diff
@@ -0,0 +1,175 @@
+#!/bin/sh
+set -e
+
+# This file is part of BOINC.
+# http://boinc.berkeley.edu
+# Copyright (C) 2018 University of California
+#
+# BOINC is free software; you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License
+# as published by the Free Software Foundation,
+# either version 3 of the License, or (at your option) any later version.
+#
+# BOINC is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with BOINC.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+export ANDROID_HOME=$HOME/Android/Sdk
+export NDK_ROOT=$HOME/Android/Ndk
+export ANDROID_TC=$HOME/Android/Toolchains
+
+# checks if a given path is canonical (absolute and does not contain relative links)
+# from http://unix.stackexchange.com/a/256437
+isPathCanonical() {
+  case "x$1" in
+    (x*/..|x*/../*|x../*|x*/.|x*/./*|x./*)
+        rc=1
+        ;;
+    (x/*)
+        rc=0
+        ;;
+    (*)
+        rc=1
+        ;;
+  esac
+  return $rc
+}
+
+doclean=""
+cache_dir=""
+arch=""
+while [ $# -gt 0 ]; do
+    key="$1"
+    case $key in
+        --cache_dir)
+        cache_dir="$2"
+        shift
+        ;;
+        --build_dir)
+        build_dir="$2"
+        shift
+        ;;
+        --clean)
+        doclean="yes"
+        ;;
+        --arch)
+        arch="$2"
+        shift
+        ;;
+        *)
+        echo "unrecognized option $key"
+        ;;
+    esac
+    shift # past argument or value
+done
+
+if [ "x$cache_dir" != "x" ]; then
+    if isPathCanonical "$cache_dir" && [ "$cache_dir" != "/" ]; then
+         PREFIX="$cache_dir"
+     else
+         echo "cache_dir must be an absolute path without ./ or ../ in it"
+         exit 1
+     fi
+else
+    cd ../
+    PREFIX="$(pwd)/3rdParty/buildCache/android-tc"
+    cd android/
+fi
+
+if [ "x$build_dir" != "x" ]; then
+    if isPathCanonical "$build_dir" && [ "$build_dir" != "/" ]; then
+         PREFIX="$build_dir"
+     else
+         echo "build_dir must be an absolute path without ./ or ../ in it"
+         exit 1
+     fi
+else
+    cd ../
+    BUILD_DIR="$(pwd)/3rdParty/android"
+    cd android/
+fi
+
+mkdir -p "${PREFIX}"
+mkdir -p "${BUILD_DIR}"
+
+if [ "${doclean}" = "yes" ]; then
+    echo "cleaning cache"
+    rm -rf "${PREFIX}"
+    mkdir -p "${PREFIX}"
+    echo "cleaning build dir"
+    rm -rf "${BUILD_DIR}"
+    mkdir -p "${BUILD_DIR}"
+fi
+
+export COMPILEOPENSSL="no"
+export OPENSSL_VERSION=1.0.2p
```

Comment:
Let's move any version details (OpenSSL, cURL, NDK) to the top of the file or, even better, refactor them into a separate config-like file in the `<root>/android` directory. Ideally that file would use a syntax that's usable by all of the build scripts (incl. `Vagrantfile`) to avoid redundant settings.

Reply:
@brevilo, It is a good idea to have config file. What exact type of config file will work for you to use it in Vagrant?

Reply:
It would basically have to be ruby, so that's out of the question I suppose since I don't see how to combine bash and ruby declarations in the same file (they share the comment operator). So just factor out version details for all the shell/bash files first of all. Maybe we'll find a way to make use of those for Vagrant as well at some later stage.

Thanks

Reply:
@brevilo,
Hm, so it sounds like a refactoring. I think, this should be done out of the scope of current task.
Will it work for you for now just to move all version declarations to the top of the file and then add a new task to think how can we move all dependencies versions to dedicated file?

Reply:
Yeah, that'd be fine.

Reply:
Have a look at https://github.com/BOINC/boinc/blob/master/mac_build/dependencyNames.sh where Charlie did something similar for the Mac build. The script gets `source`d from other build scripts and we only track the versions in one place. Maybe this can be used to create a ruby syntax on the fly with some `sed` rules?

But I also agree that this should be done afterwards.